File: openai_voice_companion.py
"""
OpenAI Voice Chat Companion with Farsi Transcription Support
This module provides a voice-enabled chat companion using OpenAI's APIs
with support for Farsi language transcription and text-to-speech.
"""

import os
import json
from pathlib import Path
from typing import Optional, List, Dict
import asyncio
from datetime import datetime

try:
    from openai import OpenAI, AsyncOpenAI
    import pyaudio
    import wave
except ImportError:
    print("Please install required packages: pip install openai pyaudio")
    raise


class FarsiVoiceChatCompanion:
    """
    A voice-enabled chat companion with Farsi support using OpenAI's API.
    Supports both speech-to-text (transcription) and text-to-speech.
    """

    def __init__(
        self,
        api_key: Optional[str] = None,
        model: str = "gpt-4",
        voice: str = "nova",
        language: str = "fa",  # Farsi language code
        sample_rate: int = 16000,
        chunk_size: int = 1024
    ):
        """
        Initialize the voice chat companion.

        Args:
            api_key: OpenAI API key (or set OPENAI_API_KEY env variable)
            model: GPT model to use for chat
            voice: Voice for TTS (alloy, echo, fable, onyx, nova, shimmer)
            language: Language code (fa for Farsi)
            sample_rate: Audio sample rate in Hz
            chunk_size: Audio chunk size for processing
        """
        self.api_key = api_key or os.getenv("OPENAI_API_KEY")
        if not self.api_key:
            raise ValueError("OpenAI API key is required")

        self.client = OpenAI(api_key=self.api_key)
        self.async_client = AsyncOpenAI(api_key=self.api_key)

        self.model = model
        self.voice = voice
        self.language = language
        self.sample_rate = sample_rate
        self.chunk_size = chunk_size

        # Conversation history
        self.conversation_history: List[Dict[str, str]] = []

        # Audio settings
        self.audio_format = pyaudio.paInt16
        self.channels = 1

    def record_audio(self, duration: int = 5, output_file: str = "input.wav") -> str:
        """
        Record audio from microphone.

        Args:
            duration: Recording duration in seconds
            output_file: Path to save the recording

        Returns:
            Path to the recorded audio file
        """
        print(f"Recording for {duration} seconds...")

        p = pyaudio.PyAudio()

        stream = p.open(
            format=self.audio_format,
            channels=self.channels,
            rate=self.sample_rate,
            input=True,
            frames_per_buffer=self.chunk_size
        )

        frames = []
        for i in range(0, int(self.sample_rate / self.chunk_size * duration)):
            data = stream.read(self.chunk_size)
            frames.append(data)

        print("Recording finished.")

        stream.stop_stream()
        stream.close()
        p.terminate()

        # Save the recording
        wf = wave.open(output_file, 'wb')
        wf.setnchannels(self.channels)
        wf.setsampwidth(p.get_sample_size(self.audio_format))
        wf.setframerate(self.sample_rate)
        wf.writeframes(b''.join(frames))
        wf.close()

        return output_file

    def transcribe_audio(self, audio_file: str) -> str:
        """
        Transcribe audio to text using OpenAI's Whisper API with Farsi support.

        Args:
            audio_file: Path to the audio file

        Returns:
            Transcribed text in Farsi
        """
        print(f"Transcribing audio from {audio_file}...")

        with open(audio_file, "rb") as audio:
            transcript = self.client.audio.transcriptions.create(
                model="whisper-1",
                file=audio,
                language=self.language,  # Farsi language
                response_format="text"
            )

        print(f"Transcription: {transcript}")
        return transcript

    def chat(self, user_message: str, system_prompt: Optional[str] = None) -> str:
        """
        Send a message to the chat model and get a response.

        Args:
            user_message: The user's message
            system_prompt: Optional system prompt for context

        Returns:
            The assistant's response
        """
        messages = []

        # Add system prompt if provided
        if system_prompt:
            messages.append({"role": "system", "content": system_prompt})
        elif not self.conversation_history:
            # Default system prompt for Farsi support
            messages.append({
                "role": "system",
                "content": "You are a helpful AI assistant that can communicate in Farsi (Persian). Respond naturally and helpfully to the user's questions."
            })

        # Add conversation history
        messages.extend(self.conversation_history)

        # Add current user message
        messages.append({"role": "user", "content": user_message})

        print(f"Sending message to {self.model}...")

        response = self.client.chat.completions.create(
            model=self.model,
            messages=messages
        )

        assistant_message = response.choices[0].message.content

        # Update conversation history
        self.conversation_history.append({"role": "user", "content": user_message})
        self.conversation_history.append({"role": "assistant", "content": assistant_message})

        print(f"Assistant: {assistant_message}")
        return assistant_message

    def text_to_speech(self, text: str, output_file: str = "output.mp3") -> str:
        """
        Convert text to speech using OpenAI's TTS API.

        Args:
            text: Text to convert to speech
            output_file: Path to save the audio file

        Returns:
            Path to the generated audio file
        """
        print(f"Converting text to speech...")

        response = self.client.audio.speech.create(
            model="tts-1",
            voice=self.voice,
            input=text
        )

        response.stream_to_file(output_file)
        print(f"Audio saved to {output_file}")

        return output_file

    def play_audio(self, audio_file: str):
        """
        Play an audio file.

        Args:
            audio_file: Path to the audio file (WAV or MP3)
        """
        try:
            import subprocess

            # Try to use system audio player
            if os.path.exists("/usr/bin/mpg123"):
                subprocess.run(["mpg123", audio_file])
            elif os.path.exists("/usr/bin/ffplay"):
                subprocess.run(["ffplay", "-nodisp", "-autoexit", audio_file])
            else:
                print(f"Audio saved to {audio_file}. Please play it with your preferred audio player.")
        except Exception as e:
            print(f"Could not play audio automatically: {e}")
            print(f"Audio saved to {audio_file}. Please play it with your preferred audio player.")

    def voice_conversation_turn(
        self,
        record_duration: int = 5,
        play_response: bool = True
    ) -> Dict[str, str]:
        """
        Execute one turn of voice conversation: record -> transcribe -> chat -> speak.

        Args:
            record_duration: Duration to record user's voice in seconds
            play_response: Whether to play the assistant's response as audio

        Returns:
            Dictionary with transcription and response
        """
        # Record user audio
        audio_file = self.record_audio(duration=record_duration)

        # Transcribe to Farsi text
        user_text = self.transcribe_audio(audio_file)

        # Get chat response
        response_text = self.chat(user_text)

        # Convert response to speech
        if play_response:
            audio_response = self.text_to_speech(response_text)
            self.play_audio(audio_response)

        return {
            "user_text": user_text,
            "assistant_text": response_text
        }

    async def async_voice_conversation_turn(
        self,
        record_duration: int = 5,
        play_response: bool = True
    ) -> Dict[str, str]:
        """
        Async version of voice conversation turn.
        """
        # Record user audio (sync operation)
        audio_file = self.record_audio(duration=record_duration)

        # Transcribe to Farsi text
        with open(audio_file, "rb") as audio:
            transcript = await self.async_client.audio.transcriptions.create(
                model="whisper-1",
                file=audio,
                language=self.language,
                response_format="text"
            )

        user_text = transcript

        # Get chat response
        messages = self.conversation_history + [{"role": "user", "content": user_text}]
        response = await self.async_client.chat.completions.create(
            model=self.model,
            messages=messages
        )

        response_text = response.choices[0].message.content

        # Update conversation history
        self.conversation_history.append({"role": "user", "content": user_text})
        self.conversation_history.append({"role": "assistant", "content": response_text})

        # Convert response to speech
        if play_response:
            audio_response = await self.async_client.audio.speech.create(
                model="tts-1",
                voice=self.voice,
                input=response_text
            )

            output_file = "output.mp3"
            audio_response.stream_to_file(output_file)
            self.play_audio(output_file)

        return {
            "user_text": user_text,
            "assistant_text": response_text
        }

    def save_conversation(self, filename: str = "conversation.json"):
        """
        Save the conversation history to a JSON file.

        Args:
            filename: Path to save the conversation
        """
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump({
                "timestamp": datetime.now().isoformat(),
                "language": self.language,
                "conversation": self.conversation_history
            }, f, ensure_ascii=False, indent=2)

        print(f"Conversation saved to {filename}")

    def load_conversation(self, filename: str):
        """
        Load a conversation history from a JSON file.

        Args:
            filename: Path to the conversation file
        """
        with open(filename, 'r', encoding='utf-8') as f:
            data = json.load(f)
            self.conversation_history = data.get("conversation", [])

        print(f"Conversation loaded from {filename}")

    def clear_conversation(self):
        """Clear the conversation history."""
        self.conversation_history = []
        print("Conversation history cleared.")


def main():
    """
    Example usage of the Farsi Voice Chat Companion.
    """
    print("=" * 60)
    print("OpenAI Farsi Voice Chat Companion")
    print("=" * 60)

    # Initialize the companion
    companion = FarsiVoiceChatCompanion(
        model="gpt-4",
        voice="nova",
        language="fa"
    )

    print("\nAvailable commands:")
    print("1. Voice conversation (record and chat)")
    print("2. Text conversation (type and chat)")
    print("3. Transcribe audio file")
    print("4. Save conversation")
    print("5. Exit")

    while True:
        print("\n" + "-" * 60)
        choice = input("\nEnter your choice (1-5): ").strip()

        if choice == "1":
            duration = input("Recording duration in seconds (default 5): ").strip()
            duration = int(duration) if duration else 5

            try:
                result = companion.voice_conversation_turn(
                    record_duration=duration,
                    play_response=True
                )
                print(f"\nYou said (Farsi): {result['user_text']}")
                print(f"Assistant: {result['assistant_text']}")
            except Exception as e:
                print(f"Error: {e}")

        elif choice == "2":
            user_input = input("\nYour message (Farsi or English): ").strip()
            if user_input:
                try:
                    response = companion.chat(user_input)

                    # Optionally convert to speech
                    speak = input("Convert response to speech? (y/n): ").strip().lower()
                    if speak == 'y':
                        audio_file = companion.text_to_speech(response)
                        companion.play_audio(audio_file)
                except Exception as e:
                    print(f"Error: {e}")

        elif choice == "3":
            audio_file = input("Enter audio file path: ").strip()
            if os.path.exists(audio_file):
                try:
                    transcript = companion.transcribe_audio(audio_file)
                    print(f"\nTranscription: {transcript}")
                except Exception as e:
                    print(f"Error: {e}")
            else:
                print("File not found!")

        elif choice == "4":
            filename = input("Save conversation to (default: conversation.json): ").strip()
            filename = filename if filename else "conversation.json"
            companion.save_conversation(filename)

        elif choice == "5":
            print("\nGoodbye!")
            break

        else:
            print("Invalid choice. Please try again.")


if __name__ == "__main__":
    main()

File: openai_voice_companion.js
/**
 * OpenAI Voice Chat Companion with Farsi Transcription Support (Node.js)
 * This module provides a voice-enabled chat companion using OpenAI's APIs
 * with support for Farsi language transcription and text-to-speech.
 */

const OpenAI = require('openai');
const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');

class FarsiVoiceChatCompanion {
  /**
   * Initialize the voice chat companion.
   *
   * @param {Object} options - Configuration options
   * @param {string} options.apiKey - OpenAI API key (or set OPENAI_API_KEY env variable)
   * @param {string} options.model - GPT model to use for chat (default: "gpt-4")
   * @param {string} options.voice - Voice for TTS (alloy, echo, fable, onyx, nova, shimmer)
   * @param {string} options.language - Language code (fa for Farsi)
   */
  constructor(options = {}) {
    this.apiKey = options.apiKey || process.env.OPENAI_API_KEY;

    if (!this.apiKey) {
      throw new Error('OpenAI API key is required');
    }

    this.client = new OpenAI({ apiKey: this.apiKey });

    this.model = options.model || 'gpt-4';
    this.voice = options.voice || 'nova';
    this.language = options.language || 'fa'; // Farsi language code

    // Conversation history
    this.conversationHistory = [];
  }

  /**
   * Transcribe audio to text using OpenAI's Whisper API with Farsi support.
   *
   * @param {string} audioFilePath - Path to the audio file
   * @returns {Promise<string>} Transcribed text in Farsi
   */
  async transcribeAudio(audioFilePath) {
    console.log(`Transcribing audio from ${audioFilePath}...`);

    const audioFile = fs.createReadStream(audioFilePath);

    const transcript = await this.client.audio.transcriptions.create({
      file: audioFile,
      model: 'whisper-1',
      language: this.language, // Farsi language
      response_format: 'text'
    });

    console.log(`Transcription: ${transcript}`);
    return transcript;
  }

  /**
   * Send a message to the chat model and get a response.
   *
   * @param {string} userMessage - The user's message
   * @param {string} systemPrompt - Optional system prompt for context
   * @returns {Promise<string>} The assistant's response
   */
  async chat(userMessage, systemPrompt = null) {
    const messages = [];

    // Add system prompt if provided
    if (systemPrompt) {
      messages.push({ role: 'system', content: systemPrompt });
    } else if (this.conversationHistory.length === 0) {
      // Default system prompt for Farsi support
      messages.push({
        role: 'system',
        content: 'You are a helpful AI assistant that can communicate in Farsi (Persian). Respond naturally and helpfully to the user\'s questions.'
      });
    }

    // Add conversation history
    messages.push(...this.conversationHistory);

    // Add current user message
    messages.push({ role: 'user', content: userMessage });

    console.log(`Sending message to ${this.model}...`);

    const response = await this.client.chat.completions.create({
      model: this.model,
      messages: messages
    });

    const assistantMessage = response.choices[0].message.content;

    // Update conversation history
    this.conversationHistory.push({ role: 'user', content: userMessage });
    this.conversationHistory.push({ role: 'assistant', content: assistantMessage });

    console.log(`Assistant: ${assistantMessage}`);
    return assistantMessage;
  }

  /**
   * Convert text to speech using OpenAI's TTS API.
   *
   * @param {string} text - Text to convert to speech
   * @param {string} outputFile - Path to save the audio file
   * @returns {Promise<string>} Path to the generated audio file
   */
  async textToSpeech(text, outputFile = 'output.mp3') {
    console.log('Converting text to speech...');

    const mp3 = await this.client.audio.speech.create({
      model: 'tts-1',
      voice: this.voice,
      input: text
    });

    const buffer = Buffer.from(await mp3.arrayBuffer());
    await fs.promises.writeFile(outputFile, buffer);

    console.log(`Audio saved to ${outputFile}`);
    return outputFile;
  }

  /**
   * Play an audio file using system audio player.
   *
   * @param {string} audioFile - Path to the audio file
   */
  async playAudio(audioFile) {
    return new Promise((resolve, reject) => {
      let player;

      // Try different audio players based on platform
      if (process.platform === 'darwin') {
        // macOS
        player = spawn('afplay', [audioFile]);
      } else if (process.platform === 'linux') {
        // Linux
        player = spawn('mpg123', [audioFile]);
      } else if (process.platform === 'win32') {
        // Windows
        player = spawn('powershell', ['-c', `(New-Object Media.SoundPlayer '${audioFile}').PlaySync()`]);
      }

      if (player) {
        player.on('close', (code) => {
          if (code === 0) {
            resolve();
          } else {
            console.log(`Audio saved to ${audioFile}. Please play it with your preferred audio player.`);
            resolve();
          }
        });

        player.on('error', (err) => {
          console.log(`Could not play audio automatically: ${err.message}`);
          console.log(`Audio saved to ${audioFile}. Please play it with your preferred audio player.`);
          resolve();
        });
      } else {
        console.log(`Audio saved to ${audioFile}. Please play it with your preferred audio player.`);
        resolve();
      }
    });
  }

  /**
   * Execute one turn of voice conversation: transcribe -> chat -> speak.
   *
   * @param {string} audioInputFile - Path to the recorded audio file
   * @param {boolean} playResponse - Whether to play the assistant's response as audio
   * @returns {Promise<Object>} Object with user_text and assistant_text
   */
  async voiceConversationTurn(audioInputFile, playResponse = true) {
    // Transcribe to Farsi text
    const userText = await this.transcribeAudio(audioInputFile);

    // Get chat response
    const responseText = await this.chat(userText);

    // Convert response to speech
    if (playResponse) {
      const audioResponse = await this.textToSpeech(responseText);
      await this.playAudio(audioResponse);
    }

    return {
      user_text: userText,
      assistant_text: responseText
    };
  }

  /**
   * Save the conversation history to a JSON file.
   *
   * @param {string} filename - Path to save the conversation
   */
  async saveConversation(filename = 'conversation.json') {
    const data = {
      timestamp: new Date().toISOString(),
      language: this.language,
      conversation: this.conversationHistory
    };

    await fs.promises.writeFile(
      filename,
      JSON.stringify(data, null, 2),
      'utf-8'
    );

    console.log(`Conversation saved to ${filename}`);
  }

  /**
   * Load a conversation history from a JSON file.
   *
   * @param {string} filename - Path to the conversation file
   */
  async loadConversation(filename) {
    const data = JSON.parse(
      await fs.promises.readFile(filename, 'utf-8')
    );

    this.conversationHistory = data.conversation || [];
    console.log(`Conversation loaded from ${filename}`);
  }

  /**
   * Clear the conversation history.
   */
  clearConversation() {
    this.conversationHistory = [];
    console.log('Conversation history cleared.');
  }
}

// Example usage
async function main() {
  console.log('='.repeat(60));
  console.log('OpenAI Farsi Voice Chat Companion (Node.js)');
  console.log('='.repeat(60));

  try {
    // Initialize the companion
    const companion = new FarsiVoiceChatCompanion({
      model: 'gpt-4',
      voice: 'nova',
      language: 'fa'
    });

    // Example 1: Text conversation
    console.log('\n--- Example 1: Text Conversation ---');
    const response1 = await companion.chat('سلام! چطوری؟');
    console.log(`Response: ${response1}`);

    // Example 2: Transcribe an audio file (you need to provide an audio file)
    // console.log('\n--- Example 2: Transcribe Audio ---');
    // const transcript = await companion.transcribeAudio('path/to/audio.mp3');
    // console.log(`Transcript: ${transcript}`);

    // Example 3: Text-to-Speech
    console.log('\n--- Example 3: Text to Speech ---');
    await companion.textToSpeech('سلام! من یک دستیار هوش مصنوعی هستم.', 'farsi_greeting.mp3');

    // Example 4: Full voice conversation (requires audio file)
    // console.log('\n--- Example 4: Voice Conversation Turn ---');
    // const result = await companion.voiceConversationTurn('input_audio.mp3', true);
    // console.log(`You said: ${result.user_text}`);
    // console.log(`Assistant: ${result.assistant_text}`);

    // Example 5: Save conversation
    console.log('\n--- Example 5: Save Conversation ---');
    await companion.saveConversation('my_conversation.json');

  } catch (error) {
    console.error('Error:', error.message);
  }
}

// Run the example if this file is executed directly
if (require.main === module) {
  main();
}

module.exports = FarsiVoiceChatCompanion;
